File: LeetCode/791.Custom_Sort_String/Mario/791.Custom_Sort_String.py
#!/usr/bin/env python3


class Solution:
    def customSortString(self, order: str, s: str) -> str:
        # counter(s)
        # based on order to permute

        count = Counter(s)

        res = ""
        for i, c in enumerate(order):
            if c in count:
                res += c * count[c]
            del count[c]

        for k, v in count.items():
            res += k * v

        return res

File: LeetCode/791.Custom_Sort_String/791.Custom_Sort_String.swift
class Solution {
    func customSortString(_ order: String, _ s: String) -> String {
        // lowercase
        var freq: [Character: Int] = [:]
        for ch in s {
            freq[ch, default: 0] += 1
        }

        var res = ""
        for ch in order {
            if let f = freq[ch] {
                res += String(repeating: String(ch), count: f)
                freq[ch] = nil
            }
        }

        for (ch, f) in freq {
            res += String(repeating: String(ch), count: f)
        }

        return res
    }
}

class Solution {
    func customSortString(_ order: String, _ s: String) -> String {
        var rank: [Character: Int] = [:]
        for (i, ch) in order.enumerated() {
            rank[ch] = i
        }

        let n = s.count
        var res = Array(s)
        res.sort { a, b -> Bool in
            let ra = rank[a] ?? n
            let rb = rank[b] ?? n
            return ra < rb
        }

        return String(res)
    }
}
